include/rapidcheck/state/Commands.hpp

```
@@ -49,5 +196,81 @@ void showValue(const std::vector<
   }
 }
 
+template <typename CommandType>
+struct ParallelCommands {
+  using Cmd = CommandType;
+  using Cmds = Commands<CommandType>;
+
+  ParallelCommands() {}
+
+  ParallelCommands(const Cmds &prefix, const Cmds &left, const Cmds &right)
+      : prefix(prefix)
+      , left(left)
+      , right(right) {}
+
+  Cmds prefix;
+  Cmds left;
+  Cmds right;
+};
+
+template <typename Cmd, typename Model, typename Sut>
+void runAllParallel(const ParallelCommands<Cmd> &commands, const Model &state, Sut &sut) {
+
+  // Verify pre-conditions for all possible interleavings
+  detail::isValidSequence(commands, state);
+
+  detail::ParallelExecutionResult<Model, Cmd> result;
+
+  // Run serial commands
+  for (const auto &command : commands.prefix) {
+    auto verifyFunc = command->run(sut);
+    result.prefix.emplace_back(
```

That's fine, but in that case, `isValidExecution` should not be called `isValidExecution` since it will throw instead of returning `bool` :)